src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs

```diff
@@ -212,71 +212,165 @@ public DynamoView(DynamoViewModel dynamoViewModel)
-                    if (contentControl != null)
-                    {
-                        tab.Content = contentControl.Content;
-                    }
-                    var extensionWindow = contentControl as Window;
-                    if (extensionWindow != null)
+                    tab = AddExtensionTab(viewExtension, contentControl);
+                }
+            }
+            else
+            {
+                // Set focus on the existing control
+                if (window != null)
+                {
+                    window.Focus();
+                }
+                else if (tab != null)
+                {
+                    // Make sure the extension bar is visible
+                    if (ExtensionsCollapsed)
                     {
-                        // Make sure the extension window closes with Dynamo
-                        extensionWindow.Owner = this;
+                        ToggleExtensionBarCollapseStatus();
                     }
+
+                    tabDynamic.SelectedItem = tab;
                 }
+            }
 
-                //Insert the tab at the end
-                ExtensionTabItems.Insert(ExtensionTabItems.Count, tab);
+            return addExtensionControl;
+        }
 
-                tabDynamic.DataContext = ExtensionTabItems;
-                tabDynamic.SelectedItem = tab;
+        private ExtensionWindow AddExtensionWindow(IViewExtension viewExtension, ContentControl contentControl, WindowSettings windowSettings)
+        {
+            ExtensionWindow window;
+            if (windowSettings == null)
+            {
+                window = new ExtensionWindow();
+                window.Owner = this;
+            }
+            else
+            {
+                var windowRect = new ModelessChildWindow.WindowRect()
+                {
+                    Left = windowSettings.Left,
+                    Top = windowSettings.Top,
+                    Width = windowSettings.Width,
+                    Height = windowSettings.Height
+                };
+                window = new ExtensionWindow(this, ref windowRect);
+                if (windowSettings.Status == WindowStatus.Maximized)
+                {
+                    // Rather than setting the WindowState here, this is delayed to the Loaded event.
+                    // This helps overcome a bug which makes the window appear always on the primary screen.
+                    window.ShouldMaximize = true;
+                }
+            }
+            
+            // Icon is passed from DynamoView (respecting Host integrator icon)
+            SetApplicationIcon();
+            window.Icon = this.Icon;
+            window.ExtensionContent.Content = contentControl;
+            window.Title = viewExtension.Name;
+            window.Tag = viewExtension;
+            window.Uid = viewExtension.UniqueId;
+            window.Closing += ExtensionWindow_Closing;
+            window.Closed += ExtensionWindow_Closed;
```

Comment:
These reminds me, do we need to unsubscribe these when docking?

Reply:
No because these windows no longer live after that.